File: api/core/hunt/ingestion/tools/factcheck.py
"""Google Fact Check Tools — claim verification.

API key REQUIRED (set ``FACTCHECKTOOLS_API_KEY``). When the key is
missing the tool returns zero findings (the runner is responsible
for filtering out the key-required stubs at registration time).
"""
from __future__ import annotations

import json
from typing import Any, List, Mapping

import httpx

from hunt.ingestion.base import Finding, ToolFunction
from hunt.config import get_settings
from hunt.models import EntityKind, SourceTool


class FactCheckTool(ToolFunction):
    name = SourceTool.FACTCHECK
    accepts = {"claim", "domain"}
    emits = {EntityKind.CLAIM, EntityKind.URL}
    key_required = True
    description = "Google Fact Check Tools — claim verification search."
    docs_url = "https://developers.google.com/fact-check/tools/api"
    per_request_timeout = 10.0

    URL = "https://factchecktools.googleapis.com/v1alpha1/claims:search"

    def _fetch(self, query: Mapping[str, Any]) -> httpx.Response:
        key = get_settings().factchecktools_api_key
        with httpx.Client(timeout=self.per_request_timeout) as client:
            return client.get(
                self.URL,
                params={"query": query["target"], "key": key, "pageSize": 20},
            )

    def _parse(self, raw: bytes, query: Mapping[str, Any]) -> List[Finding]:
        try:
            data = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            return []
        findings: List[Finding] = []
        for claim in data.get("claims", []):
            text = claim.get("text", "")
            claim_url = claim.get("claimReview", [{}])[0].get("url") if claim.get("claimReview") else None
            publisher = (
                claim.get("claimReview", [{}])[0].get("publisher", {}).get("name")
                if claim.get("claimReview")
                else None
            )
            rating = (
                claim.get("claimReview", [{}])[0].get("textualRating")
                if claim.get("claimReview")
                else None
            )
            if text:
                findings.append(
                    Finding(
                        source_tool=self.name,
                        entity_kind=EntityKind.CLAIM,
                        entity_value=text,
                        attributes={
                            "publisher": publisher,
                            "rating": rating,
                            "url": claim_url,
                        },
                    )
                )
        return findings

File: lib/tools/factcheck.ts
/**
 * Google Fact Check Tools — claim verification.
 *
 * API key required (FACTCHECKTOOLS_API_KEY). The registry hides
 * the tool when the key is missing.
 */

import { ToolFunction, type FetchResult, type Finding } from "./base";
import { getSettings } from "../config";

export class FactCheckTool extends ToolFunction {
  readonly name = "factcheck";
  readonly accepts = new Set(["claim", "domain", "person", "org"]);
  readonly emits = new Set(["claim", "url"]);
  readonly key_required = true;
  readonly docs_url = "https://developers.google.com/fact-check/tools/api";
  readonly description = "Google Fact Check Tools — claim verification search.";
  readonly per_request_timeout_ms = 10_000;

  private static URL = "https://factchecktools.googleapis.com/v1alpha1/claims:search";

  protected async _fetch(query: { target: string }): Promise<FetchResult> {
    const key = getSettings().factchecktools_api_key;
    const url = new URL(FactCheckTool.URL);
    url.searchParams.set("query", query.target);
    url.searchParams.set("key", key || "");
    url.searchParams.set("pageSize", "20");
    const r = await fetch(url.toString(), {
      signal: AbortSignal.timeout(this.per_request_timeout_ms),
    });
    return {
      bytes: Buffer.from(await r.arrayBuffer()),
      contentType: r.headers.get("content-type") || "application/json",
    };
  }

  protected _parse(raw: Buffer, _query: { target: string }): Finding[] {
    let data: any = {};
    try {
      data = JSON.parse(raw.toString("utf-8"));
    } catch {
      return [];
    }
    const findings: Finding[] = [];
    for (const claim of data.claims || []) {
      const text = claim.text || "";
      const reviews = claim.claimReview || [];
      const first = reviews[0] || {};
      if (text) {
        findings.push({
          source_tool: this.name,
          entity_kind: "claim",
          entity_value: text,
          attributes: {
            publisher: first.publisher?.name,
            rating: first.textualRating,
            url: first.url,
          },
        });
      }
    }
    return findings;
  }
}
